url0307c0e995_js_labs_js_collider_tgz-pJ8-org_jsl_tests_shmem_throughput_MainJ8/src/org/jsl/collider/PerfCounter.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (C) 2013 Sergey Zubarev, info@js-labs.org
+ *
+ * This file is a part of JS-Collider framework.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.jsl.collider;
+
+import java.util.concurrent.atomic.AtomicLongArray;
+
+public class PerfCounter
+{
+    /* 0 - min
+     * 1 - avg
+     * 2 - max
+     * 3 - sum
+     */
+    private final String m_name;
+    private final AtomicLongArray m_data;
+
+    public PerfCounter( String name )
+    {
+        m_name = name;
+        m_data = new AtomicLongArray(4);
+        m_data.set( 0, Long.MAX_VALUE );
+    }
+
+    public final void trace( long startTime )
+    {
+        final long delay = (System.nanoTime() - startTime);
+
+        for (;;)
+        {
+            long min = m_data.get(0);
+            if (delay >= min)
+                break;
+            if (m_data.compareAndSet(0, min, delay))
+                break;
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 45:52, 67:74.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.